src/pseudo/PspUpf.jl

```diff
@@ -128,101 +120,104 @@ function PspUpf(path; identifier=path)
+    end
 
-function PspUpf(Zion, lmax, rgrid::Vector{T}, drgrid, vloc, r_projs, h, pswfcs, pswfc_occs,
-                r2_4π_ρion, ρcore; identifier="", description="") where {T <: Real}
-    vloc_interp = linear_interpolation((rgrid, ), vloc)
+    vloc_interp = linear_interpolation((rgrid,), vloc)
     r_projs_interp = map(r_projs) do r_projs_l
         map(r_projs_l) do r_proj  # Can't use views here; have to match `vloc_interp`'s type
             ir_cut = lastindex(r_proj)
-            linear_interpolation((rgrid[1:ir_cut], ), r_proj)
+            linear_interpolation((rgrid[1:ir_cut],), r_proj)
         end
     end
-    r2_4π_ρion_interp = linear_interpolation((rgrid, ), r2_4π_ρion)
-    ρcore_interp = linear_interpolation((rgrid, ), ρcore)
+    r2_4π_ρion_interp = linear_interpolation((rgrid,), r2_4π_ρion)
+    ρcore_interp = linear_interpolation((rgrid,), ρcore)
 
-    r_vloc_corr_dr = (rgrid .* vloc .+ Zion) .* drgrid
-    r2_projs_dr = map(r_projs) do r_projs_l
-        @views map(r_projs_l) do r_proj
-            ir_cut = lastindex(r_proj)
-            rgrid[1:ir_cut] .* r_proj .* drgrid[1:ir_cut]
-        end
-    end
-    r2_4π_ρion_dr = r2_4π_ρion .* drgrid
-    r2_ρcore_dr = rgrid.^2 .* ρcore .* drgrid
-
-    PspUpf{T,typeof(vloc_interp)}(Zion, lmax, rgrid, drgrid, vloc, r_projs, h, pswfcs,
-                                  pswfc_occs, r2_4π_ρion, ρcore, vloc_interp, r_projs_interp,
-                                  r2_4π_ρion_interp, ρcore_interp, r_vloc_corr_dr,
-                                  r2_projs_dr, r2_4π_ρion_dr, r2_ρcore_dr,
-                                  identifier, description)
+    PspUpf{eltype(rgrid),typeof(vloc_interp)}(Zion, lmax, rgrid, drgrid, vloc,
+        r_projs, h, pswfcs, pswfc_occs, r2_4π_ρion, ρcore, vloc_interp,
+        r_projs_interp, r2_4π_ρion_interp, ρcore_interp, identifier, description
+    )
 end
 
 charge_ionic(psp::PspUpf) = psp.Zion
 has_valence_density(psp::PspUpf) = !all(iszero, psp.r2_4π_ρion)
 has_core_density(psp::PspUpf) = !all(iszero, psp.ρcore)
 
 # Note: UPFs store `r[j] β_{li}(r[j])`, so r=0 is undefined and will error.
-function eval_psp_projector_real(psp::PspUpf, i, l, r::T)::T where {T <: Real}
+function eval_psp_projector_real(psp::PspUpf, i, l, r::T)::T where {T<:Real}
     psp.r_projs_interp[l+1][i](r) / r
 end
 
 # For UPFs, the integral is transformed to the following sum:
 # 4π Σ{k} j_l(q r[k]) (r[k]^2 p_{il}(r[k]) dr[k])
-function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
-    r2_proj_dr = psp.r2_projs_dr[l+1][i]
-    s = zero(T)
-    @inbounds for ir = eachindex(r2_proj_dr)
-        s += sphericalbesselj_fast(l, q * psp.rgrid[ir]) * r2_proj_dr[ir]
-    end
-    4T(π) * s
+function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T<:Real}
+    ir_cut = length(psp.r_projs[l+1][i])
+    f = (
+        psp.rgrid[1:ir_cut] .* psp.r_projs[l+1][i] .*
+        sphericalbesselj_fast.(l, q .* psp.rgrid[1:ir_cut])
```

Comment:
For these things I like putting the `.*` in front to make the operation more visible

Reply:
You also probably want `@views`. I sometimes wonder if we should not enclose the whole of DFTK in an `@views`